docs/DesignDocs/AttestationAPIsWithFlexibleClaims.md

```diff
@@ -0,0 +1,816 @@
+Attestation APIs with Init- and Run-Time Claims Support
+==========================
+
+Motivation
+------------
+
+The Open Enclave SDK currently supports attestation APIs,
+`oe_get_attestation_certificate_with_evidence_v2` and
+`oe_verify_attestation_certificate_with_evidence_v2`.
+The `oe_get_attestation_certificate_with_evidence_v2` API
+allows a developer to obtain a self-signed certificate with
+the enclave evidence embedded.
+The `oe_verify_attestation_certificate_with_evidence_v2`
+allows a developer to validate the certificate and
+to verify the embedded enclave evidence (based on the DCAP client).
+In addition, the API outputs a list of claims, allowing the
+developer to implement additional verification logic.
+These APIs are useful to implement an end-to-end
+attested TLS handshake where one party sends out the certificate generated by
+another party verifies the enclave evidence from the certificate based on the
+`oe_verify_attestation_certificate_with_evidence_v2` API
+(see the [attested TLS sample](/samples/attested_tls) for more detail).
+However, these APIs have the following limitations:
+
+- Fail to support general use cases
+
+  A typical attestation procedure includes three roles: an attester, a relying party,
+  and a verifier. Two general models are as follows.
+  - Passport Model
+
+    The attester generates the evidence and asks the verifier to assess the
+    trustworthiness of the evidence. Then, the attester sends the attestation results
+    to the relying party. Finally, the relying party validates the attestation results
+    based on customized policies.
+
+  - Background-Check Model
+
+    The attester generates the evidence and directly sends the evidence to the
+    relying party. The relying party then asks the verifier to assess the trustworthiness
+    of the evidence and then validates the attestation results based on customized policies.
+
+  The end-to-end use case supported by the APIs, however, assumes that the party
+  that sends the certificate is the attester and the receiver of the certificate plays
+  the roles of both the verifier and the relying party. More specifically, the
```

Comment:
The existing API has been designed to specifically support the background check mode, with a combined verifier + relying-party entity. The design in this document feels like a patch to stretch the existing API to support the new usages (especially the passport mode). I think it would be a better approach to start a with a top-down design to build a layer on top of the existing API. For passport mode support, I'd like to see a detailed description of the usages, the cert + attestation result design, and the end-to-end flows between the three parties (attester, verifier, relying party).

Reply:
I agree to have a different APIs other than extending the existing APIs. For example, simply `oe_get_attestation_certificate` and `oe_parse_attestation_certificate`. Please refer the user experience part for the example of work flow.

The biggest concern I have is how much backward compatibility the new APIs need to provide. Also, given that we have the `oe_create_attestation_ceritifcate` API already, we would need a better naming for the new get API.

Reply:
Note: this PR could be seen as an effort to design a higher-level API to is more aligned to Dave's document [Proposed Attestation API Requirements](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/AttestationApiRequirements.md). The existing API can be seen as a lower-level one to build on top.

Reply:
First generation - `oe_generate_attestation_certificate` embeds OE_REPORT in OID 1.3.6.1.4.1.311.105.1
Second generation - `oe_get_attestation_certificate_with_evidence*` embeds OE_EVIDENCE in OID 1.3.6.1.4.1.311.105.2

Report and evidence have subtle difference.

We might consider this new thing embedded in certificate an extension of evidence and give it a new OID (e.g. 1.3.6.1.4.1.311.105.2.1). This will guarantee backward compatibility. We just have to be careful picking OID to avoid collision or misinterpretation.